mirai-core/src/commonMain/kotlin/network/notice/priv/FriendGroupNoticeProcessor.kt

```diff
@@ -0,0 +1,116 @@
+) : MixedNoticeProcessor(), NewContactSupport {
+
+    override suspend fun NoticePipelineContext.processImpl(data: MsgType0x210) = data.context {
+        when (data.uSubMsgType) {
+            0x27L -> {
+                val body = vProtobuf.loadAs(Submsgtype0x27.SubMsgType0x27.SubMsgType0x27MsgBody.serializer())
+                for (msgModInfo in body.msgModInfos) {
+                    markAsConsumed(msgModInfo)
+                    when {
+                        msgModInfo.msgModFriendGroup != null -> handleFriendGroupChanged(
+                            msgModInfo.msgModFriendGroup, logger
+                        )
+                        msgModInfo.msgModGroupName != null -> handleFriendGroupNameChanged(
+                            msgModInfo.msgModGroupName, logger
+                        )
+                        msgModInfo.msgDelGroup != null -> handleDelGroup(msgModInfo.msgDelGroup, logger)
+                        msgModInfo.msgAddGroup != null -> handleAddGroup(msgModInfo.msgAddGroup)
+                        else -> markNotConsumed(msgModInfo)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun NoticePipelineContext.handleAddGroup(
+        addGroup: Submsgtype0x27.SubMsgType0x27.AddGroup
+    ) {
+        bot.friendGroups.friendGroups.add(
+            FriendGroupImpl(
+                bot,
+                FriendGroupInfo(addGroup.groupid, String(addGroup.groupname), 0, 0)
+            )
+        )
+    }
+
+    private fun NoticePipelineContext.handleDelGroup(
+        delGroup: Submsgtype0x27.SubMsgType0x27.DelGroup, logger: MiraiLogger
+    ) {
+        bot.friendGroups[delGroup.groupid]?.let { friendGroup ->
+            val defaultFriendGroup = bot.friendGroups[0]!!
+            friendGroup.friends.forEach {
+                it.impl().info.friendGroupId = 0
+                defaultFriendGroup.friends.delegate.add(it)
+            }
+            bot.friendGroups.friendGroups.remove(friendGroup)
+        } ?: let {
+            logger.error { "fail to find FriendGroup(id=${delGroup.groupid}) in Bot(id=${bot.id})" }
+            return
+        }
+    }
+
+    private fun NoticePipelineContext.handleFriendGroupNameChanged(
+        modFriendGroup: Submsgtype0x27.SubMsgType0x27.ModGroupName, logger: MiraiLogger
+    ) {
+        bot.friendGroups[modFriendGroup.groupid]?.let {
+            it.impl().name = String(modFriendGroup.groupname)
+        } ?: let {
+            logger.error { "fail to find FriendGroup(id=${modFriendGroup.groupid}) in Bot(id=${bot.id})" }
+            return
+        }
+    }
+
+    private fun NoticePipelineContext.handleFriendGroupChanged(
+        modFriendGroup: Submsgtype0x27.SubMsgType0x27.ModFriendGroup,
+        logger: MiraiLogger
+    ) {
+        modFriendGroup.msgFrdGroup.forEach { body ->
+            val friend = bot.getFriend(body.fuin) ?: let {
+                logger.error { "fail to find Friend(id=${body.fuin}) in Bot(id=${bot.id})" }
```

Comment:
[ERROR] Detected friend XXX was moved to friend group XXX but friend not found in bot's friends list